### PR TITLE
Delay all enactments by one extra epoch

### LINKED ISSUE #3691: Postpone Enactment of all GovActions to next epoch
According to [CIP-1694](https://github.com/cardano-foundation/CIPs/tree/master/CIP-1694#lifecycle) `HardForkInitiation` and `ParameterChange` governance actions should be enacted a full epoch after ratification.

In order to avoid complex changes to the spec and implementation we decided to postpone enactment for one more epoch. This is necessary:
* for proper functioning HardFork combinator in consensus (2 stability windows)
* for block header validation (1 stability window)

In the end we postpone enactment for 2 epochs:
* 1 epoch to compute stake distributions
* 1 more epoch for allowing for safe consensus

In the future we will be able optimize it to one 1 epoch, but implementation will have to be a lot more complex

### PR DESCRIPTION
**Pull request by Soupstraw**
# Description

Delay all enactments by one extra epoch. The order of events for a governance action from proposal to enactment looks like this:

```
-- Epoch 0 --
Action P is proposed

<wait some epochs until P gets enough votes or expires>

-- Start of epoch n --
Take a snapshot and start calculating DRep distributions
  (we also do that on all previous epochs, but here we assume that this time the action is either ratified or expires)

-- Start of epoch n+1 --
If P has enough votes:
  - Apply P onto `cgFutureEnactState`
Remove P from the snapshots

-- Start of epoch n+2 --
Rotate `cgFutureEnactState` to `cgEnactState`
```

closes #3691

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

**Comment by teodanciu** [comment]
Thanks for writing the example! 

In the implementation, where is this part? 

```If P has enough votes:
  - Apply P onto `cgFutureEnactState`
Remove P from the snapshots
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs [hunk 1]
```diff
@@ -551,10 +558,11 @@ instance EraPParams era => ToJSON (ConwayGovState era) where
 instance EraPParams era => ToExpr (ConwayGovState era)
 
 toConwayGovPairs :: (KeyValue a, EraPParams era) => ConwayGovState era -> [a]
-toConwayGovPairs cg@(ConwayGovState _ _) =
+toConwayGovPairs cg@(ConwayGovState _ _ _) =
   let ConwayGovState {..} = cg
    in [ "gov" .= cgGovSnapshots
-      , "ratify" .= cgEnactState
+      , "curEnact" .= cgEnactState
+      , "futureEnact" .= cgEnactState
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs]:
```suggestion
      , "futureEnactState" .= cgFutureEnactState
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs [hunk 2]
```diff
@@ -551,10 +558,11 @@ instance EraPParams era => ToJSON (ConwayGovState era) where
 instance EraPParams era => ToExpr (ConwayGovState era)
 
 toConwayGovPairs :: (KeyValue a, EraPParams era) => ConwayGovState era -> [a]
-toConwayGovPairs cg@(ConwayGovState _ _) =
+toConwayGovPairs cg@(ConwayGovState _ _ _) =
   let ConwayGovState {..} = cg
    in [ "gov" .= cgGovSnapshots
-      , "ratify" .= cgEnactState
+      , "curEnact" .= cgEnactState
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Governance.hs]: Nice catch on `ratify`. Let's give it a more complete name
```suggestion
      , "curEnactState" .= cgEnactState
```

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs [hunk 3]
```diff
@@ -242,13 +243,14 @@ epochTransition = do
...
     ratSig = RatifySignal . snapshotActions . prevGovSnapshots $ cgGovSnapshots govSt
+    futureEnactState = govSt ^. cgFutureEnactStateL
   RatifyState {rsRemoved, rsEnactState} <-
     trans @(EraRule "RATIFY" era) $
       TRC
         ( ratEnv
         , RatifyState
             { rsRemoved = mempty
-            , rsEnactState = govSt ^. cgEnactStateL
+            , rsEnactState = futureEnactState
```
Comment (lehins) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs]: Are you sure this is suppose to be `futureEncastState`? I could of course be wrong, but in my understanding it should be `govSt ^. cgEnactStateL`.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/ConwayFeatures.hs [hunk 4]
```diff
@@ -426,25 +436,25 @@ testGov pf = do
...
 
   assertExprEqualWithMessage "govState after second proposal" govState2 expectedGovState2
 
-  -- Wait two epochs
+  -- Wait three epochs
+  let
+    epochState2 = epochState1'' & esLStateL .~ ledgerState3
+  epochState3 <- expectRight "Error running runEPOCH: " $ runEPOCH (EPOCH pf) epochState2 (EpochNo 4) poolDistr
+  epochState3' <- expectRight "Error running runEPOCH: " $ runEPOCH (EPOCH pf) epochState3 (EpochNo 4) poolDistr
+  epochState4 <- expectRight "Error running runEPOCH: " $ runEPOCH (EPOCH pf) epochState3' (EpochNo 5) poolDistr
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/ConwayFeatures.hs]: This looks slightly wrong, why do we run EPOCH rule on the same epoch number, but different state? Also we should stick to one convention of either ticks or numbers, without mixing them in the same function:
```suggestion
  epochState4 <- expectRight "Error running runEPOCH: " $ runEPOCH (EPOCH pf) epochState3 (EpochNo 5) poolDistr
  epochState5 <- expectRight "Error running runEPOCH: " $ runEPOCH (EPOCH pf) epochState4 (EpochNo 6) poolDistr
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/ConwayFeatures.hs [hunk 5]
```diff
@@ -399,19 +403,25 @@ testGov pf = do
...
             ]
         )
-    -- Wait two epochs
+    -- Wait three epochs
     eitherEpochState1 = runEPOCH (EPOCH pf) epochState0 (EpochNo 2) poolDistr
   epochState1 <- expectRight "Error running runEPOCH: " eitherEpochState1
   let
     constitution = epochState1 ^. esLStateL . lsUTxOStateL . utxosGovStateL . cgEnactStateL . ensConstitutionL
   assertExprEqualWithMessage "constitution after one epoch" constitution def
+  epochState1' <-
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/ConwayFeatures.hs]: Same here could you avoid mixing ticks with numbers. It is hard to follow the dependency order

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/ConwayFeatures.hs [hunk 6]
```diff
@@ -399,19 +403,25 @@ testGov pf = do
...
+    expectRight "Error running runEPOCH: " $
+      runEPOCH (EPOCH pf) epochState1 (EpochNo 3) poolDistr
+  let
+    constitution' = epochState1' ^. esLStateL . lsUTxOStateL . utxosGovStateL . cgEnactStateL . ensConstitutionL
+  assertExprEqualWithMessage "constitution two epochs" constitution' def
+  epochState1'' <-
+    expectRight "Error running runEPOCH: " $
+      runEPOCH (EPOCH pf) epochState1' (EpochNo 3) poolDistr
```
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Examples/ConwayFeatures.hs]: Again same EpochNo as in the previous call to runEPOCH

**Review by lehins — CHANGES_REQUESTED** [review]
I am very suspicious of the EnactState argument to the RATIFY rule. We should do some more analysis and more tests in order to ensure we got it right.

**Review by WhatisRT — COMMENTED** [review]
I'm not sure I can provide a meaningful review, as I've got no idea how the delay for the incremental computation works. To me the code that didn't get changed looks completely wrong already, since all the inputs to `RATIFY` need to be snapshotted (if they aren't known statically), but it looks to me like the ingredients of `ratEnv` are just computed (transitively) from `es`.

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs (hunk 3 above)
Comment (Soupstraw) [reply eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs]: I think it should be `futureEnactState`, because that's where we stored the enactments last epoch. Now one epoch later this becomes the actual enact state.

**Comment by lehins** [comment]
> since all the inputs to RATIFY need to be snapshotted

@WhatisRT They are snapshotted. Stake distribution for Pools and DReps is from the previous epoch. We just need to make sure that the starting EnactState for the RATIFY rules also comes from the previous epoch

**Comment by Soupstraw** [comment]
@teodanciu 
> In the implementation, where is this part?
> 
> ```
>   - Apply P onto `cgFutureEnactState`
> Remove P from the snapshots
> ```

This is when we run `RATIFY` using `futureEnactState`

File: eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs [hunk 7]
```diff
@@ -242,13 +243,14 @@ epochTransition = do
         , reDRepState = vstate ^. vsDRepsL
         }
     ratSig = RatifySignal . snapshotActions . prevGovSnapshots $ cgGovSnapshots govSt
+    futureEnactState = govSt ^. cgFutureEnactStateL
```
Comment (teodanciu) [inline eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Epoch.hs]: As we said in the meeting, `enactState` would be a less confusing name than `futureEnactState`

**Comment by lehins** [comment]
This can be closed, as we decided not use this approach and instead we'll run RATIFY rule two stability windows before the end of the epoch (implemented in #3759), thus allowing us to postpone enactment only for one epoch, rather than two.
